File: M3Tools/Controls/GenericInputPair.vb
Imports System.ComponentModel

Public Class GenericInputPair
	Public Shadows Event TextChanged As EventHandler

	<DefaultValue("Label1")>
	Public Property LabelText As String
		Get
			Return lbl_InputLabel.Text
		End Get
		Set(value As String)
			lbl_InputLabel.Text = value
		End Set
	End Property

	Public ReadOnly Property TextLength As Integer
		Get
			Return txt_Input.TextLength
		End Get
	End Property

	Public Property Mask As String
		Get
			Return txt_Input.Mask
		End Get
		Set(value As String)
			txt_Input.Mask = value
		End Set
	End Property

	<DefaultValue(False)>
	Public Property UseSystemPasswordChar As Boolean
		Get
			Return txt_Input.UseSystemPasswordChar
		End Get
		Set(value As Boolean)
			txt_Input.UseSystemPasswordChar = value
		End Set
	End Property

	Public Overrides Property Text As String
		Get
			Return If(txt_Input.Text <> Placeholder, txt_Input.Text, "")
		End Get
		Set(value As String)
			txt_Input.Text = If(value <> Placeholder AndAlso value <> "", value, Placeholder)
		End Set
	End Property

	Public Property TextAlign As Windows.Forms.HorizontalAlignment
		Get
			Return txt_Input.TextAlign
		End Get
		Set(value As Windows.Forms.HorizontalAlignment)
			txt_Input.TextAlign = value
		End Set
	End Property

	<DefaultValue(False)>
	Public Property [ReadOnly] As Boolean
		Get
			Return txt_Input.ReadOnly
		End Get
		Set(value As Boolean)
			txt_Input.ReadOnly = value
		End Set
	End Property

	<DefaultValue("Input...")>
	Public Property Placeholder As String = "Input..."

	Private Sub InputTextChanged(sender As Object, e As EventArgs) Handles txt_Input.TextChanged
		' TODO: Potentially move placeholder/color logic here?

		RaiseEvent TextChanged(Me, e)
	End Sub

	Private Sub InputGotFocus(sender As Object, e As EventArgs) Handles txt_Input.GotFocus
		If txt_Input.Text <> Placeholder Then
			Return
		End If

		txt_Input.Text = ""
		txt_Input.ForeColor = Drawing.SystemColors.WindowText
	End Sub

	Private Sub InputLostFocus(sender As Object, e As EventArgs) Handles txt_Input.LostFocus
		If txt_Input.Text <> "" Then
			Return
		End If

		txt_Input.Text = Placeholder
		txt_Input.ForeColor = Drawing.SystemColors.ControlDark
	End Sub

	Private Sub GenericInputPair_Load(sender As Object, e As EventArgs) Handles Me.Load
		Text = If(Text <> "", Text, Placeholder)
		txt_Input.ForeColor = If(Text <> "", Drawing.SystemColors.WindowText, Drawing.SystemColors.ControlDark)
	End Sub
End Class

File: M3Tools/Controls/GenericInputPair.cs
using System;
using System.ComponentModel;

namespace SPPBC.M3Tools
{
	/// <summary>
	/// 
	/// </summary>
    public partial class GenericInputPair
    {
		/// <summary>
		/// 
		/// </summary>
        public new event EventHandler TextChanged;

		/// <summary>
		/// The text that should be shown in the label
		/// </summary>
        [DefaultValue("Label1")]
        [RefreshProperties(RefreshProperties.Repaint)]
        public string Label
        {
            get
            {
                return lbl_InputLabel.Text;
            }
            set
            {
                lbl_InputLabel.Text = value;
            }
        }

		/// <summary>
		/// The length of the text in the textbox
		/// </summary>
        public int TextLength
        {
            get
            {
                return txt_Input.TextLength;
            }
        }

		/// <summary>
		/// The mask that should be applied to the textbox
		/// </summary>
        public string Mask
        {
            get
            {
                return txt_Input.Mask;
            }
            set
            {
                txt_Input.Mask = value;
            }
        }

		/// <summary>
		/// Whether to use the password character to obfuscate the text
		/// </summary>
        [DefaultValue(false)]
        public bool UseSystemPasswordChar
        {
            get
            {
                return txt_Input.UseSystemPasswordChar;
            }
            set
            {
                txt_Input.UseSystemPasswordChar = value;
            }
        }

		/// <summary>
		/// The text within the textbox
		/// </summary>
        [RefreshProperties(RefreshProperties.Repaint)]
        public override string Text
        {
            get
            {
                return txt_Input.Text != Placeholder ? txt_Input.Text : "";
            }
            set
            {
				if (string.IsNullOrEmpty(value) || value == Placeholder)
				{
					txt_Input.Text = Placeholder;
					return;
				}

                txt_Input.Text = value;
            }
        }

		/// <summary>
		/// How the text in the textbox should be aligned
		/// </summary>
        [RefreshProperties(RefreshProperties.Repaint)]
        public System.Windows.Forms.HorizontalAlignment TextAlign
        {
            get
            {
                return txt_Input.TextAlign;
            }
            set
            {
                txt_Input.TextAlign = value;
                Refresh();
            }
        }

		/// <summary>
		/// Whether the textbox is readonly
		/// </summary>
        [DefaultValue(false)]
        public bool ReadOnly
        {
            get
            {
                return txt_Input.ReadOnly;
            }
            set
            {
                txt_Input.ReadOnly = value;
            }
        }

		/// <summary>
		/// The placeholder that should be shown in the textbox when empty
		/// </summary>
        [DefaultValue("Input...")]
        [RefreshProperties(RefreshProperties.Repaint)]
        public string Placeholder { get; set; }

		/// <summary>
		/// 
		/// </summary>
        public GenericInputPair()
        {
            InitializeComponent();
        }

        private void InputTextChanged(object sender, EventArgs e)
        {
            // TODO: Potentially move placeholder/color logic here?

            TextChanged?.Invoke(this, e);
        }

        private void InputGotFocus(object sender, EventArgs e)
        {
            if (txt_Input.Text != Placeholder)
            {
                txt_Input.Select(txt_Input.Text.Length, 0);
                txt_Input.ScrollToCaret();
                return;
            }

            txt_Input.Text = "";
            txt_Input.ForeColor = System.Drawing.SystemColors.WindowText;
        }

        private void InputLostFocus(object sender, EventArgs e)
        {
            if (!string.IsNullOrEmpty(txt_Input.Text))
            {
                return;
            }

            txt_Input.Text = Placeholder;
            txt_Input.ForeColor = System.Drawing.SystemColors.ControlDark;
        }

        private void Loading(object sender, EventArgs e)
        {
            Text = !string.IsNullOrEmpty(Text) ? Text : Placeholder;
            txt_Input.ForeColor = !string.IsNullOrEmpty(Text) ? System.Drawing.SystemColors.WindowText : System.Drawing.SystemColors.ControlDark;
        }
    }
}
